<a href="https://colab.research.google.com/github/Narrayane/02_Aprendendo_a_fazer_ETL_Alura/blob/main/02_02_Pandas_Trabalhando_diferentes_formatos_de_arquivos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
url = 'https://raw.githubusercontent.com/alura-cursos/Pandas/refs/heads/main/superstore_data.csv'

In [ ]:
dados_mercado = pd.read_csv(url)

In [ ]:
dados_mercado.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [ ]:
url2 = 'https://raw.githubusercontent.com/alura-cursos/Pandas/refs/heads/main/superstore_data_ponto_virgula.csv'

In [ ]:
dados_ponto_virgula = pd.read_csv(url2)

In [ ]:
dados_ponto_virgula.head()

Id;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;Response;Complain
0  1826;1970;Graduation;Divorced;84835;0;0;6/16/2...                                                                                                                                                                                                                             
1  1;1961;Graduation;Single;57091;0;0;6/15/2014;0...                                                                                                                                                                                                                             
2  10476;1958;Graduation;Married;67267;0;1;5/13/2...                                                                                                                                                                                                                             
3  1386;1967;Graduation;Together;32474;1;1;11/5/2...                                                                                                                                                                                                                             
4  5371;1989;Graduation;Single;21474;1;0;8/4/2014...

In [ ]:
#para ler um arquivo separaco com por ponto e vírgula utilize o parâmetro abaixo:
dados_ponto_virgula = pd.read_csv(url2, sep=';')
dados_ponto_virgula.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [ ]:
# ver as primeiras 5 linhas utilizo o parâmetro nrows conforme abaixo
dados_primeiras_linhas = pd.read_csv(url, nrows=5)
dados_primeiras_linhas.head()

Id  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced   84835        0         0   
1      1        1961  Graduation         Single   57091        0         0   
2  10476        1958  Graduation        Married   67267        0         1   
3   1386        1967  Graduation       Together   32474        1         1   
4   5371        1989  Graduation         Single   21474        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [ ]:
# ver algumas colunas utilizo o parâmetro usecols coloque o = e abra colchetes para informar quais colunas quero que seja mostrada conforme abaixo
dados_selecao = pd.read_csv(url, usecols=['Id','Year_Birth','Income'])
dados_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [ ]:
# também posso colocar o índice da coluna
dados_selecao = pd.read_csv(url, usecols=[0,1,4])
dados_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [ ]:
# para salvar um arquivo em CSV utilize o parâmetro to_csv e coloque entre parentes e aspas o nome que desejo colocar no arquivo que será gerado
dados_selecao.to_csv('clientes_mercado.csv')

In [ ]:
# Para ler o arquivo gerado:
clientes_mercado = pd.read_csv('/content/clientes_mercado.csv')
clientes_mercado

Unnamed: 0     Id  Year_Birth   Income
0              0   1826        1970  84835.0
1              1      1        1961  57091.0
2              2  10476        1958  67267.0
3              3   1386        1967  32474.0
4              4   5371        1989  21474.0
...          ...    ...         ...      ...
2235        2235  10142        1976  66476.0
2236        2236   5263        1977  31056.0
2237        2237     22        1976  46310.0
2238        2238    528        1978  65819.0
2239        2239   4070        1969  94871.0

[2240 rows x 4 columns]

In [ ]:
# houve uma duplicação do index. Para isso podemos usar o parâmetro index que por padrão é True para criar o índice. Como não queremos que crie um índice iremos informar False
dados_selecao.to_csv('clientes_mercado.csv', index = False)
dados_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

Giovanna é uma cientista de dados que trabalha em uma empresa de comércio eletrônico. Ela é responsável por analisar os dados de vendas para identificar padrões e tendências que possam ajudar a empresa a tomar decisões informadas.

Recentemente, ela recebeu uma nova tarefa: analisar um arquivo CSV contendo informações sobre as vendas da empresa em um determinado período. Para isso, Giovanna tentou fazer a leitura do um arquivo CSV no Google Colab, utilizando a biblioteca Pandas com o seguinte comando:

import pandas as pd
df = pd.read_csv('/content/dados.csv')
Porém, surgiu o erro abaixo:
Em português a mensagem quer dizer UnicodeDecodeError: o codec 'utf-8' não pode decodificar o byte 0xe7 na posição 18: byte de continuação inválido

Mas o que isso quer dizer? O que é esse tal de UTF-8?
O erro de encoding ocorre quando a biblioteca Pandas não consegue interpretar corretamente os caracteres de um arquivo CSV. Isso pode acontecer quando ele contém caracteres especiais que não são reconhecidos pela biblioteca Pandas ou quando foi salvo em um formato de codificação diferente do esperado.

Para resolver esse erro, é necessário identificar a codificação correta do arquivo CSV e especificá-la ao carregar o arquivo com a biblioteca Pandas. Essa codificação padrão é o UTF-8, mas em alguns casos, o arquivo pode ter sido salvo com uma codificação diferente, como ISO-8859-1.

O UTF-8 é uma codificação de caracteres universal usada para representar caracteres de diferentes idiomas de forma compatível com a internet e com sistemas de computador em geral. A sigla UTF significa Unicode Transformation Format (Formato de Transformação Unicode) e o número 8 indica que essa codificação associa uma sequência de 1 a 4 bytes (8 a 32 bits) com cada caractere.

A codificação UTF-8 é amplamente utilizada na internet e em sistemas de computador em todo o mundo, pois permite a representação de caracteres de diferentes idiomas em um único conjunto de caracteres. Além disso, essa codificação é capaz de preservar a compatibilidade com outras mais antigas, como ASCII, o que a torna uma escolha popular para a criação e compartilhamento de arquivos de texto.

Pode ser que você esteja pensando neste momento: como será que a Giovanna conseguiu descobrir qual é a codificação do arquivo que ela está tentando ler?

Existem algumas formas para descobrir isso, no entanto, teremos a oportunidade de experimentar uma maneira prática de fazer isso no próprio Google Colab. Vamos lá?

Nós podemos usar uma biblioteca chamada chardet para detectar o encoding de um arquivo CSV. Para utilizar essa biblioteca no Google Colab, basta realizar sua importação:

import chardet

Em seguida, digitamos o seguinte bloco de código:

with open('/content/dados.csv', 'rb') as file:
    print(chardet.detect(file.read()))

    Note que a primeira linha deste código abre o arquivo CSV em modo de leitura binária rb e atribui o objeto de arquivo retornado a uma variável chamada file.

Na segunda linha do código, o conteúdo do arquivo CSV é lido usando o método read() e o resultado é passado para a função chardet.detect(), que retorna um dicionário contendo informações sobre a codificação mais provável do arquivo. O resultado é impresso com a função print() que mostra a codificação e a confiança associada a essa codificação.

Ao executar o código Giovanna obteve o seguinte resultado:

{'encoding': 'ISO-8859-1', 'confidence': 0.73, 'language': ''}

Agora ela sabe que o arquivo CSV provavelmente possui a codificação ISO-8859-1 com uma confiança de 0.73. Para especificar a codificação correta ao carregar o arquivo CSV com a biblioteca Pandas, é possível usar um parâmetro chamado encoding:

df = pd.read_csv('/content/dados.csv', encoding='ISO-8859-1')

Desta forma, a biblioteca Pandas irá carregar o arquivo CSV usando a codificação ISO-8859-1, o que vai resolver o erro de encoding. Então, se algum dia você tentar ler um arquivo e surgir esse mesmo erro, lembre-se dessa dica!

A biblioteca Pandas é uma das ferramentas mais poderosas disponíveis para a análise de dados em Python. Ela oferece uma ampla variedade de funcionalidades que tornam o processo de manipulação e análise de dados muito mais fácil e eficiente.

Para aproveitar ao máximo todas as funcionalidades da biblioteca, é essencial explorar a sua documentação, que é extensa e bem organizada, e pode ser facilmente acessada no modo on-line. Confira as indicações:

Função read_csv = https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

Função to_csv = https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

Explorar a documentação da biblioteca é importante por diversas razões. Ela pode ajudar a entender como as funções e métodos da biblioteca funcionam e como elas podem ser usadas para analisar e manipular os dados.

Além disso, a documentação fornece detalhes sobre os parâmetros de cada função e exemplos de como eles podem ser usados na prática. Então, busque aprofundar os seus estudos nesses parâmetros das funções read_csv e to_csv e aprenda mais sobre cada um deles.

**Aula 2 - Utilizando planilhas **

Dados: https://github.com/alura-cursos/Pandas/blob/main/emissoes_CO2.xlsx

In [ ]:
import pandas as pd

In [ ]:
url = 'https://github.com/alura-cursos/Pandas/blob/main/emissoes_CO2.xlsx?raw=True'
# para trazer o arquivo bruto é necessário acrescentar o ?raw=True

In [ ]:
dados_co2 = pd.read_excel(url)

In [ ]:
dados_co2.head()

País ISO 3166-1 alpha-3   Ano  Total  Carvão  Oléo  Gás  Cimento  \
0  Afeganistão                AFG  1750    0.0     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751    0.0     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752    0.0     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753    0.0     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754    0.0     NaN   NaN  NaN      NaN   

   Queima  Outro  Per Capita  
0     NaN    NaN         NaN  
1     NaN    NaN         NaN  
2     NaN    NaN         NaN  
3     NaN    NaN         NaN  
4     NaN    NaN         NaN

In [ ]:
# Uma planilha poderá ter várias abas ou páginas. Para descobrir utilizo o seguinte método:
pd.ExcelFile(url).sheet_names

#neste exemplo temos três páginas ou abas

['emissoes_C02', 'emissoes_percapita', 'fontes']

** 03
Para saber mais: aprofundando na função read_excel **

O Microsoft Excel é uma das ferramentas de planilha mais utilizadas no mundo, sendo amplamente utilizado para armazenar e analisar dados em formato de tabela. No vídeo anterior, demonstramos como a função read_excel pode ser utilizada para ler um arquivo Excel no formato xlsx. No entanto, essa função é capaz de ler arquivos em outros formatos, como: xls, xlsm, xlsb, odf, ods e odt.

Vamos entender um pouco mais sobre esses diferentes tipos de arquivos?

xls
O xls é um formato de arquivo do Excel mais antigo e usado até a versão de 2003.

xlsx
Já o xlsx é o formato de arquivo padrão do Excel a partir da versão 2007. Esse formato é baseado em XML (Extensible Markup Language - Linguagem de marcação extensível) e é amplamente suportado por outras ferramentas de planilha eletrônica, incluindo o Google Planilhas.

Atenção: Se surgiram algumas dúvidas com o termo XML, não se preocupe! Mais adiante teremos uma aula sobre esse tema. Nela você entenderá o que é esse formato, na medida em que lê e escreve em XML.

xlsm
Há também o xlsm, uma extensão de arquivo usada pelo Excel para armazenar planilhas que contêm as macros. Estas são sequências de comandos ou instruções que podem ser executadas automaticamente para realizar tarefas específicas na planilha.

Então, o formato xlsm permite que as macros sejam salvas juntamente com a planilha, para que possam ser executadas sempre que a planilha for aberta.

xlsb
Por fim, temos o formato xlsb, uma extensão de arquivo usada pelo Excel para armazenar planilhas em formato binário. A codificação binária permite que as planilhas sejam abertas e salvas com mais rapidez do que aquelas no formato xlsx.

Já os formatos odf, ods e odt são arquivos abertos, livres e universais que podem ser usados por qualquer software, ou seja, foram criados para serem independentes de plataformas, isso significa que podem ser usados em sistemas operacionais diferentes, como Windows, Linux e Mac OS.

Além disso, são independentes de aplicativos e podem ser usados em vários programas diferentes, incluindo o OpenOffice, o LibreOffice, o Google Docs e o Microsoft Office. Esse padrão de arquivos foi criado e é mantido pela OASIS (Organization for the Advancement of Structured Information Standards), uma organização internacional criada para desenvolver e promover padrões digitais para uso na Internet.

Legal! Agora sabemos um pouco mais sobre todos os tipos de arquivos que podem ser lidos com a funçãoread_excel.

Caso queira aprofundar o assunto, deixamos alguns links para materiais que foram utilizados como referência:

https://support.microsoft.com/pt-br/office/formatos-de-arquivos-com-suporte-no-excel-0943ff2c-6014-4e8d-aaea-b83d51d46247

https://pt.wikipedia.org/wiki/OpenDocument

In [ ]:
# a função sheet por padrão 0, ou seja a primeira aba ou página da planilha. Se colocarmos a 1 será a segunda aba ou página, ou podemos colocar o nome da página
percapita = pd.read_excel(url, sheet_name = 'emissoes_percapita')
percapita.head()

País ISO 3166-1 alpha-3   Ano  Total  Carvão  Óleo  Gás  Cimento  \
0  Afeganistão                AFG  1750    NaN     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751    NaN     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752    NaN     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753    NaN     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754    NaN     NaN   NaN  NaN      NaN   

   Queima  Outro  
0     NaN    NaN  
1     NaN    NaN  
2     NaN    NaN  
3     NaN    NaN  
4     NaN    NaN

In [ ]:
fontes = pd.read_excel(url, sheet_name = 'fontes')
fontes.head()

País ISO 3166-1 alpha-3   Ano   Total  Carvão    Oléo     Gás  \
0  Afeganistão                AFG  1750  [NONE]  [NONE]  [NONE]  [NONE]   
1  Afeganistão                AFG  1751  [NONE]  [NONE]  [NONE]  [NONE]   
2  Afeganistão                AFG  1752  [NONE]  [NONE]  [NONE]  [NONE]   
3  Afeganistão                AFG  1753  [NONE]  [NONE]  [NONE]  [NONE]   
4  Afeganistão                AFG  1754  [NONE]  [NONE]  [NONE]  [NONE]   

  Cimento  Queima   Outro Per Capita  
0  [NONE]  [NONE]  [NONE]     [NONE]  
1  [NONE]  [NONE]  [NONE]     [NONE]  
2  [NONE]  [NONE]  [NONE]     [NONE]  
3  [NONE]  [NONE]  [NONE]     [NONE]  
4  [NONE]  [NONE]  [NONE]     [NONE]

In [ ]:
# especificar colunas para os casos em que queremos apenas um determinado intervalo
intervalo = pd.read_excel(url, sheet_name = 'emissoes_C02', usecols='A:D')
intervalo

País ISO 3166-1 alpha-3   Ano         Total
0      Afeganistão                AFG  1750  0.000000e+00
1      Afeganistão                AFG  1751  0.000000e+00
2      Afeganistão                AFG  1752  0.000000e+00
3      Afeganistão                AFG  1753  0.000000e+00
4      Afeganistão                AFG  1754  0.000000e+00
...            ...                ...   ...           ...
63099       Global                WLD  2017  3.609674e+10
63100       Global                WLD  2018  3.682651e+10
63101       Global                WLD  2019  3.708256e+10
63102       Global                WLD  2020  3.526409e+10
63103       Global                WLD  2021  3.712385e+10

[63104 rows x 4 columns]

In [ ]:
# Com o nrows eu defico a quantidade de linhas
intervalo_2 = pd.read_excel(url, sheet_name = 'emissoes_C02', usecols='A:D', nrows=10)
intervalo_2

País ISO 3166-1 alpha-3   Ano  Total
0  Afeganistão                AFG  1750      0
1  Afeganistão                AFG  1751      0
2  Afeganistão                AFG  1752      0
3  Afeganistão                AFG  1753      0
4  Afeganistão                AFG  1754      0
5  Afeganistão                AFG  1755      0
6  Afeganistão                AFG  1756      0
7  Afeganistão                AFG  1757      0
8  Afeganistão                AFG  1758      0
9  Afeganistão                AFG  1759      0

In [ ]:
# para salvar um arquivo em excel
percapita.to_excel('co2_percapita.xlsx', index = False)

In [ ]:
pd.read_excel('/content/co2_percapita.xlsx')

País ISO 3166-1 alpha-3   Ano      Total     Carvão       Óleo  \
0      Afeganistão                AFG  1750        NaN        NaN        NaN   
1      Afeganistão                AFG  1751        NaN        NaN        NaN   
2      Afeganistão                AFG  1752        NaN        NaN        NaN   
3      Afeganistão                AFG  1753        NaN        NaN        NaN   
4      Afeganistão                AFG  1754        NaN        NaN        NaN   
...            ...                ...   ...        ...        ...        ...   
63099       Global                WLD  2017  4749682.0  1908857.0  1610910.0   
63100       Global                WLD  2018  4792753.0  1919213.0  1596350.0   
63101       Global                WLD  2019  4775633.0  1896468.0  1589920.0   
63102       Global                WLD  2020  4497423.0  1807760.0  1427353.0   
63103       Global                WLD  2021  4693699.0  1893923.0  1496614.0   

             Gás   Cimento   Queima    Outro  
0            NaN       NaN      NaN      NaN  
1            NaN       NaN      NaN      NaN  
2            NaN       NaN      NaN      NaN  
3            NaN       NaN      NaN      NaN  
4            NaN       NaN      NaN      NaN  
...          ...       ...      ...      ...  
63099   940144.0  198416.0  51579.0  39776.0  
63100   979965.0  204225.0  53634.0  39366.0  
63101   984878.0  208309.0  56569.0  39490.0  
63102   963695.0  208844.0  51981.0  37789.0  
63103  1001585.0  211472.0  52663.0  37443.0  

[63104 rows x 10 columns]

** 07 Para saber mais: importando no Google Planilhas **

O Google Planilhas é uma ferramenta de planilha eletrônica baseada em nuvem que permite criar, editar e colaborar em planilhas de forma fácil e conveniente. É um dos aplicativos do Google Workspace que oferece uma alternativa ao Microsoft Excel.

Com o Google Planilhas é possível criar e formatar planilhas, adicionar fórmulas e funções, criar gráficos e tabelas dinâmicas, além de trabalhar com dados em tempo real com outras pessoas. Ele também suporta a importação e exportação de vários formatos de arquivos, como CSV e XLSX.

Uma das principais vantagens do Google Planilhas é a colaboração em tempo real. Várias pessoas podem trabalhar simultaneamente na mesma planilha, visualizando as alterações simultaneamente. Isso torna mais fácil trabalhar em projetos em equipe, aumentando a eficiência e a produtividade.

Para importar o arquivo Excel emissoes_CO2.xlsx no Google Planilhas, siga as seguintes etapas:

Abra a página do Google Planilhas e clique em “Acesse o Sheets”. Faça login em sua conta do Google, se necessário.
Clique em “Arquivo” (File) no menu superior e depois clique na opção “Importar arquivo” (Import).
Clique na guia "Fazer upload" (Upload) e selecione o arquivo que deseja importar. Você pode clicar na opção Browse para localizar o arquivo no seu computador ou arrastar o arquivo dentro da guia.
Aguarde enquanto o arquivo é carregado. Quando o upload estiver completo, você verá uma mensagem de confirmação e pode clicar em “Importar dados” (Import data).
Após a importação os dados serão exibidos.
Maravilha! Os dados foram importados com sucesso. No próximo vídeo você aprenderá como ler esses dados no Google Colab, utilizando o link de compartilhamento do Google Planilhas.

In [ ]:
## https://docs.google.com/spreadsheets/d/1if5CPayPaUXOBRz5S0yCVBd4wWwjFXaeC-QcHQi_Pfk/edit?usp=sharing

In [ ]:
# para separar o id da planilha eu seleciono o que está entre /d/ e /edit
sheet_id = '1if5CPayPaUXOBRz5S0yCVBd4wWwjFXaeC-QcHQi_Pfk'

In [ ]:
# criar a variável usando o f-string para faciliar o entendimento
url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'

# /gviz/tq = permite que o sistema se comunique com o google planilhas ao google colab
# ?tqx=out:csv = fará com que os dados sejam retonandos como CSV
# &sheet = fará com que todos os dados sejam retornados

#copio do link copiado até o /d/, acrescento a váriável criada, em seguida o seguinte comando: /gviz/tq?tqx=out:csv&sheet

In [ ]:
dados_co2_sheets = pd.read_csv(url)
dados_co2_sheets.head()


País ISO 3166-1 alpha-3   Ano Total Carvão Oléo  Gás Cimento Queima  \
0  Afeganistão                AFG  1750     0    NaN  NaN  NaN     NaN    NaN   
1  Afeganistão                AFG  1751     0    NaN  NaN  NaN     NaN    NaN   
2  Afeganistão                AFG  1752     0    NaN  NaN  NaN     NaN    NaN   
3  Afeganistão                AFG  1753     0    NaN  NaN  NaN     NaN    NaN   
4  Afeganistão                AFG  1754     0    NaN  NaN  NaN     NaN    NaN   

  Outro Per Capita  
0   NaN        NaN  
1   NaN        NaN  
2   NaN        NaN  
3   NaN        NaN  
4   NaN        NaN

In [ ]:
# como a planilha possui três páginas iremos definir algumas informações

In [ ]:
sheet_id = '1if5CPayPaUXOBRz5S0yCVBd4wWwjFXaeC-QcHQi_Pfk'
sheet_name = 'emissoes_percapita'
url_percapita = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'

In [ ]:
percapita_sheets = pd.read_csv(url_percapita)
percapita_sheets.head()

País ISO 3166-1 alpha-3   Ano Total Carvão Óleo  Gás Cimento Queima  \
0  Afeganistão                AFG  1750   NaN    NaN  NaN  NaN     NaN    NaN   
1  Afeganistão                AFG  1751   NaN    NaN  NaN  NaN     NaN    NaN   
2  Afeganistão                AFG  1752   NaN    NaN  NaN  NaN     NaN    NaN   
3  Afeganistão                AFG  1753   NaN    NaN  NaN  NaN     NaN    NaN   
4  Afeganistão                AFG  1754   NaN    NaN  NaN  NaN     NaN    NaN   

   Outro  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

In [ ]:
sheet_id = '1if5CPayPaUXOBRz5S0yCVBd4wWwjFXaeC-QcHQi_Pfk'
sheet_name = 'fontes'
url_fontes= f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'
fontes_sheets = pd.read_csv(url_fontes)
fontes_sheets.head()

País ISO 3166-1 alpha-3   Ano   Total  Carvão    Oléo     Gás  \
0  Afeganistão                AFG  1750  [NONE]  [NONE]  [NONE]  [NONE]   
1  Afeganistão                AFG  1751  [NONE]  [NONE]  [NONE]  [NONE]   
2  Afeganistão                AFG  1752  [NONE]  [NONE]  [NONE]  [NONE]   
3  Afeganistão                AFG  1753  [NONE]  [NONE]  [NONE]  [NONE]   
4  Afeganistão                AFG  1754  [NONE]  [NONE]  [NONE]  [NONE]   

  Cimento  Queima   Outro Per Capita  
0  [NONE]  [NONE]  [NONE]     [NONE]  
1  [NONE]  [NONE]  [NONE]     [NONE]  
2  [NONE]  [NONE]  [NONE]     [NONE]  
3  [NONE]  [NONE]  [NONE]     [NONE]  
4  [NONE]  [NONE]  [NONE]     [NONE]

** Aula 03 Manipulando arquivos JSON**

In [ ]:
import pandas as pd

In [ ]:
dados_pacientes = pd.read_json('/content/pacientes.json')
dados_pacientes

ID_paciente Doenca_cardiaca    IMC Fumante Consumo_alcool  AVC  \
0              0             Nao  16.60     Sim            Nao  Nao   
1              1             Nao  20.34     Nao            Nao  Sim   
2              2             Nao  26.58     Sim            Nao  Nao   
3              3             Nao  24.21     Nao            Nao  Nao   
4              4             Nao  23.71     Nao            Nao  Nao   
..           ...             ...    ...     ...            ...  ...   
995          995             Nao  28.29     Sim            Nao  Nao   
996          996             Nao  25.69     Sim            Nao  Nao   
997          997             Sim  26.99     Nao            Nao  Nao   
998          998             Nao  44.29     Nao            Nao  Nao   
999          999             Nao  25.83     Nao            Nao  Nao   

     Saude_fisica  Saude_mental Dificuldade_caminhar Sexo_biologico  \
0               3            30                  Nao       Feminino   
1               0             0                  Nao       Feminino   
2              20            30                  Nao      Masculino   
3               0             0                  Nao       Feminino   
4              28             0                  Sim       Feminino   
..            ...           ...                  ...            ...   
995            10            20                  Nao       Feminino   
996             0             0                  Nao       Feminino   
997             1             0                  Nao      Masculino   
998            30             0                  Sim       Feminino   
999             0             0                  Nao      Masculino   

    Faixa_etaria    Raca Diabetes Atividade_fisica Saude_geral  Horas_sono  \
0          55-59  Branca      Sim              Sim   Muito boa           5   
1        80 ou +  Branca      Nao              Sim   Muito boa           7   
2          65-69  Branca      Sim              Sim    Razoavel           8   
3          75-79  Branca      Nao               No         Boa           6   
4          40-44  Branca      Nao              Sim   Muito boa           8   
..           ...     ...      ...              ...         ...         ...   
995        65-69   Preta      Sim              Sim    Razoavel           8   
996        75-79  Branca      Nao              Sim   Excelente           8   
997        70-74  Branca      Nao              Sim   Muito boa           8   
998        70-74  Branca      Sim               No         Boa           6   
999        70-74   Preta      Sim               No   Excelente           7   

    Asma Doenca_renal Cancer_pele  
0    Sim          Nao         Sim  
1    Nao          Nao         Nao  
2    Sim          Nao         Nao  
3    Nao          Nao         Sim  
4    Nao          Nao         Nao  
..   ...          ...         ...  
995  Nao          Nao         Nao  
996  Nao          Nao         Nao  
997  Nao          Nao         Nao  
998  Sim          Nao         Nao  
999  Nao          Nao         Nao  

[1000 rows x 19 columns]

In [ ]:
## exemplo de quando as informações estão aninhadas no arquivo json:
dados_pacientes_2 = pd.read_json('/content/pacientes_2.json')
dados_pacientes_2

Pesquisa   Ano  \
0  Principais Indicadores de Doenca Cardiaca  2020   
1  Principais Indicadores de Doenca Cardiaca  2020   
2  Principais Indicadores de Doenca Cardiaca  2020   

                                           Pacientes  
0  {'ID': '01', 'Faixa_etaria': '55-59', 'Sexo_bi...  
1  {'ID': '02', 'Faixa_etaria': '80 ou +', 'Sexo_...  
2  {'ID': '03', 'Faixa_etaria': '65-69', 'Sexo_bi...

In [ ]:
# Para normlaizar no dataframe um arquivo JSON  aninhado utilizamos o método json_normalize
df_normalizado = pd.json_normalize(dados_pacientes_2['Pacientes'])
df_normalizado

ID Faixa_etaria Sexo_biologico    Raça    IMC Fumante Consumo_alcool  \
0  01        55-59       Feminino  Branca  16.60     Sim            Nao   
1  02      80 ou +       Feminino  Branca  20.34     Nao            Nao   
2  03        65-69      Masculino  Branca  26.58     Sim            Nao   

   Saude_fisica  Saude_mental Dificuldade_caminhar Atividade_fisica  \
0             3            30                  Nao              Sim   
1             0             0                  Nao              Sim   
2            20            30                  Nao              Sim   

  Saude_geral  Horas_sono                Problemas_saude  
0   Muito boa           5  [Diabetes, Asma, Cancer_pele]  
1   Muito boa           7                          [AVC]  
2   Muito boa           8               [diabetes, Asma]

** 05 Para saber mais: aprofundando na normalização **
Normalização de dados é um processo importante em ciência de dados que tem como objetivo organizar e padronizar dados para facilitar a análise e comparação entre eles. Quando se trata de dados no formato JSON (JavaScript Object Notation - Notação de Objetos JavaScript), é comum que eles estejam aninhados, o que pode dificultar sua análise e manipulação.

A biblioteca Pandas possui uma função chamada json_normalize() que permite a transformação dos dados em um formato tabular, facilitando a visualização e análise das informações. A seguir vamos aprender como usar essa função para normalizar diferentes tipos de JSON em DataFrames.

Normalizando um JSON simples
Abaixo temos uma variável chamada dados e dentro dela há um objeto JSON com três chaves e seus respectivos valores:

dados = {'Pesquisa': 'Principais Indicadores de Doenca Cardiaca', 'Ano': 2020, 'Numero_Pacientes':3}
Para normalizar essa variável, podemos passa-lá dentro da função json_normalize e analisar o DataFrame obtido:

df = pd.json_normalize(dados)
df
O resultado parece ótimo! Temos 3 colunas que são as nossas chaves e uma linha contendo os valores para cada chave.


**Normalizando um JSON com vários níveis**

Existem situações em que o arquivo JSON pode conter mais de um objeto, como é o caso do exemplo abaixo, no qual temos uma lista, armazenada na variável json_lista, contendo dois objetos JSON:

json_lista = [
    { 'ID': '01', 'Faixa_etaria': '55-59', 'Sexo_biologico': 'feminino'},
    { 'ID': '02', 'Faixa_etaria': '80 ou +', 'Sexo_biologico': 'feminino'}
]


Para normalizar essa lista podemos aplicar a função json_normalize:

pd.json_normalize(json_lista)


A função json_normalize() é capaz de converter cada registro da lista em uma linha de forma tabular.



**Normalizando um JSON com uma lista aninhada**

Bom, notamos que a função json_normalize() funciona muito bem nas situações anteriores, mas e em outras situações?

Dados como um dicionário
Vamos começar analisando a normalização quando os dados são um dicionário. Temos um dicionário armazenado na variável json_obj. Note que na chave “Saude” temos outro dicionário:

json_obj = {
    'ID': '01',
    'Faixa_etaria': '55-59',
    'Sexo_biologico': 'Feminino',
    'Saude': {'Dificuldade_caminhar': 'Nao',
              'Atividade_fisica': 'Sim',
              'IMC': 16.6,
              'Doenca_cardiaca': 'Nao',
          }
      }

Fazendo a normalização:

pd.json_normalize(json_obj)

O DataFrame gerado possui uma coluna para cada informação contida no dicionário que estava em “Saude”. As colunas criadas possuem o prefixo “Saude.”, pois as informações vieram dessa mesma chave.

**Dados como uma lista de dicionários**

Agora nós temos uma lista de dicionários armazenada na variável json_list:

json_list = [
    {
    'ID': '01',
    'Faixa_etaria': '55-59',
    'Sexo_biologico': 'Feminino',
    'Saude': {'Dificuldade_caminhar': 'Nao',
              'Atividade_fisica': 'Sim',
              'IMC': 16.6,
              'Doenca_cardiaca': 'Nao',
          }
      },
      {
          'ID': '02',
          'Faixa_etaria': '80 ou +',
          'Sexo_biologico': 'Feminino',
          'Saude': {'Dificuldade_caminhar': 'Nao',
                    'Atividade_fisica': 'Sim',
                    'IMC': 20.34,
                    'Doenca_cardiaca': 'Sim'}
       }
       ]

Fazendo a normalização:

pd.json_normalize(json_list)

Podemos analisar que todos os valores aninhados em cada registro da lista foram convertidos em colunas separadas. E os dados que nós normalizamos no vídeo anterior? Lembra que a normalização deles foi feita de um jeito diferente?

Recapitulando, vamos copiar os dados do arquivo pacientes_2.json e armazená-los numa variável chamada dados_dict.

dados_dict = {
  "Pesquisa": "Principais Indicadores de Doenca Cardiaca",
  "Ano": 2020,
  "Pacientes": [
    {
      "ID": "01",
      "Faixa_etaria": "55-59",
      "Sexo_biologico": "Feminino",
      "Raça": "Branca",
      "IMC": 16.6,
      "Fumante": "Sim",
      "Consumo_alcool": "Nao",
      "Saude_fisica": 3,
      "Saude_mental": 30,
      "Dificuldade_caminhar": "Nao",
      "Atividade_fisica": "Sim",
      "Saude_geral": "Muito boa",
      "Horas_sono": 5,
      "Problemas_saude": [
        "Diabetes",
        "Asma",
        "Cancer_pele"
      ]
    },
    {
      "ID": "02",
      "Faixa_etaria": "80 ou +",
      "Sexo_biologico": "Feminino",
      "Raça": "Branca",
      "IMC": 20.34,
      "Fumante": "Nao",
      "Consumo_alcool": "Nao",
      "Saude_fisica": 0,
      "Saude_mental": 0,
      "Dificuldade_caminhar": "Nao",
      "Atividade_fisica": "Sim",
      "Saude_geral": "Muito boa",
      "Horas_sono": 7,
      "Problemas_saude": [
        "AVC"
      ]
    },
    {
      "ID": "03",
      "Faixa_etaria": "65-69",
      "Sexo_biologico": "Masculino",
      "Raça": "Branca",
      "IMC": 26.58,
      "Fumante": "Sim",
      "Consumo_alcool": "Nao",
      "Saude_fisica": 20,
      "Saude_mental": 30,
      "Dificuldade_caminhar": "Nao",
      "Atividade_fisica": "Sim",
      "Saude_geral": "Muito boa",
      "Horas_sono": 8,
      "Problemas_saude": [
        "diabetes",
        "Asma"
      ]
    }
  ]
}

Se tentarmos normalizar esses dados:

pd.json_normalize(dados_dict)

Podemos observar que nossa lista aninhada é colocada em uma única coluna Pacientes. Então, nós usamos o seguinte código para normalizar os dados, especificando qual coluna está aninhada:

pd.json_normalize(dados_dict['Pacientes'])
Também podemos fazer isso utilizando o parâmetro record_path como ['Pacientes']. Esse parâmetro é usado na função pd.json_normalize() para especificar o caminho para os registros que devem ser normalizados em um DataFrame separado:

pd.json_normalize(dados_dict, record_path=['Pacientes'])

Com ambos os códigos o resultado é o mesmo.

O resultado parece ótimo, mas não inclui as colunas “Pesquisa” e “Ano”. Para incluí-las, podemos usar o parâmetro meta para especificar outras colunas que queremos no DataFrame.

pd.json_normalize(
    dados_dict,
    record_path =['Pacientes'],
    meta=['Pesquisa', 'Ano']
)

Dessa forma temos todas as colunas presentes no DataFrame!

Importante: Na aula nós realizamos a normalização em um arquivo com o formato JSON. No entanto, a função json_normalize() aceita apenas um dicionário ou uma lista de dicionários. Por isso, no vídeo foi usada a estratégia de usar o código: pd.json_normalize(dados_pacientes_2['Pacientes']). Porém, se tentarmos usar parâmetros da função json_normalize em um arquivo JSON podem surgir erros. Para contornar isso, precisamos importar o módulo json e ler os arquivos conforme o código abaixo:

#Importando a biblioteca Pandas
import pandas as pd

#Importando o módulo JSON
import json

#Lendo o arquivo json usando o módulo Python JSON
with open('pacientes_2.json','r') as f:
    dados = json.loads(f.read())

#Normalizando os dados com os parâmetros record_path e meta
pd.json_normalize(dados, record_path='Pacientes', meta=['Pesquisa', 'Ano'])

E, assim, aprendemos como normalizar arquivos JSON simples, com múltiplos níveis e aninhados.

In [ ]:
# para salvar o arquivo normalizado emJSON
df_normalizado.to_json('historico_pacientes_normalizado.json')
pd.read_json('historico_pacientes_normalizado.json')

# Podemos salvar em outros formatos como csv ou excel

ID Faixa_etaria Sexo_biologico    Raça    IMC Fumante Consumo_alcool  \
0   1        55-59       Feminino  Branca  16.60     Sim            Nao   
1   2      80 ou +       Feminino  Branca  20.34     Nao            Nao   
2   3        65-69      Masculino  Branca  26.58     Sim            Nao   

   Saude_fisica  Saude_mental Dificuldade_caminhar Atividade_fisica  \
0             3            30                  Nao              Sim   
1             0             0                  Nao              Sim   
2            20            30                  Nao              Sim   

  Saude_geral  Horas_sono                Problemas_saude  
0   Muito boa           5  [Diabetes, Asma, Cancer_pele]  
1   Muito boa           7                          [AVC]  
2   Muito boa           8               [diabetes, Asma]

**07 Para saber mais: obtendo arquivos JSON de APIs**

Para obter dados no formato JSON podemos utilizar uma API (Application Programming Interface). Ela é uma interface de programação de aplicação que permite a comunicação entre diferentes programas ou sistemas.

Atenção: Caso queira aprofundar no assunto, temos um Alura + explicando o que é uma API.

Existem vários formatos que podem ser usados em APIs e um dos mais comuns é o JSON (JavaScript Object Notation). O JSON é um formato leve de troca de informações entre sistemas que utiliza uma notação baseada em objetos do JavaScript.

Nós temos a seguinte situação-problema: “Uma empresa de tecnologia está em fase de expansão e deseja otimizar sua base de contatos e a forma como gerencia suas interações com os clientes e parceiros. A equipe de vendas e atendimento ao cliente está enfrentando dificuldades para acompanhar todos os contatos, entender as necessidades dos diferentes perfis de clientes, e personalizar o atendimento de forma eficiente. Para solucionar isso, a empresa precisa de uma maneira organizada e automatizada para coletar, gerenciar e segmentar informações de seus clientes e parceiros, como nome, e-mail, endereço, telefone e empresa em que trabalham.”

Nesta situação, podemos acessar a API JSONPlaceholder para coletar os dados dos contatos, que incluem nome, e-mail, endereço, telefone, entre outros. Com esses dados, a equipe pode melhorar suas operações e a comunicação com os clientes.

Uma maneira comum de acessar APIs é por meio da biblioteca requests. Para isso, também é necessário importar o módulo json:

import requests
import json

Os dados da JSONPlaceholder serão obtidos com a função request.get('url'). Dentro dela nós passamos o endereço da API:

dados_usuarios = requests.get('https://jsonplaceholder.typicode.com/users')

Conseguimos recuperar os resultados usando a função json.loads(). Para isso, é necessário passar a variável dados_usuarios, criada anteriormente com o atributo text que retorna o conteúdo da resposta.

resultado = json.loads(dados_usuarios.text)

Em seguida, podemos visualizar o DataFrame.

pd.DataFrame(resultado)

index	id	name	username	email	address	phone	website	company

Agora sabemos como utilizar a biblioteca requests e o módulo JSON para obter dados de APIs!

**Aula 04 Lendo páginas Web**


In [ ]:
# salve a página em Html para não perder as informações
import pandas as pd

In [ ]:
# Função para chamar um html
dados_html = pd.read_html('/content/filmes_wikipedia.html')
dados_html

[       0                      1
 0   1998             100 Movies
 1   1999              100 Stars
 2   2000             100 Laughs
 3   2001            100 Thrills
 4   2002           100 Passions
 5   2003  100 Heroes & Villains
 6   2004              100 Songs
 7   2005       100 Movie Quotes
 8   2005              25 Scores
 9   2006             100 Cheers
 10  2006            25 Musicals
 11  2007   100 Movies (Updated)
 12  2008        AFI's 10 Top 10
 13   vte                    vte,
                             Film  Release year              Director  \
 0                   Citizen Kane          1941          Orson Welles   
 1                     Casablanca          1942        Michael Curtiz   
 2                  The Godfather          1972  Francis Ford Coppola   
 3             Gone with the Wind          1939        Victor Fleming   
 4             Lawrence of Arabia          1962            David Lean   
 ..                           ...           ...                   

In [ ]:
type(dados_html) #descobrir o tipo de dados

list

In [ ]:
len(dados_html) # descobrir a quantidade de dataframe

3

In [ ]:
#cessar um dataframe específico, por exemplo o segundo
top_filmes = dados_html[1]
top_filmes

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

HTML (Hypertext Markup Language - Linguagem de marcação de hipertexto) é o código que usamos para estruturar uma página web e seu conteúdo. O HTML consiste em uma série de elementos que são usados para delimitar ou agrupar diferentes partes do conteúdo para que ele apareça ou atue de determinada maneira.

Um documento HTML começa com a tag <html> e termina com a tag </html>. Ele também contém um cabeçalho (<head>) que descreve o título da página e outras informações relevantes; e um corpo (<body>) que contém o conteúdo da página. O cabeçalho e o corpo são delimitados pelas tags <head> e </head>, e <body> e </body>, respectivamente.

Dentro do corpo podemos usar vários elementos para criar a estrutura do conteúdo. Note cada detalhe no exemplo a seguir:

tags <h1> a <h6> : criar títulos e subtítulos.
tags <p>: criar parágrafos.
tags <ul> e <li> : criar listas não ordenadas.
tags <img>: inserir imagens.
tags <table>, <tr>, <th> e <td>: criar tabelas.
Além disso, podemos usar atributos nos elementos HTML para especificar propriedades adicionais. Por exemplo: podemos usar o atributo src na tag <img> para indicar o endereço da imagem que queremos mostrar, ou o atributo href na tag <a> para indicar o endereço do link que queremos criar.

Caso queira aprofundar ainda mais nesse assunto recomendamos os seguintes artigos:

O que é o HTML e suas tags? Parte 1: estrutura básica: https://www.alura.com.br/artigos/o-que-e-html-suas-tags-parte-1-estrutura-basica

O que é o HTML e suas tags? Parte 2: elementos inline: https://www.alura.com.br/artigos/o-que-e-html-suas-tags-parte-2-elementos-inline?utm_source=gnarus&utm_medium=timeline

O que é o HTML e suas tags? Parte 3: elementos block-level: https://www.alura.com.br/artigos/html-tags-elementos-block-level?utm_source=gnarus&utm_medium=timeline

In [ ]:
# Para salvar em formato html

top_filmes.to_html('top_filmes')

In [ ]:
# Para salvar em formato csv
top_filmes.to_csv('top_filmes_1998.csv', index=False)


In [ ]:
pd.read_csv('/content/top_filmes_1998.csv')

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

In [ ]:
# xml = linguagem de marcação acessível
import pandas as pd

In [ ]:
dados_imdb = pd.read_xml('/content/imdb_top_1000.xml')
dados_imdb.head(3)

index                                        Poster_Link  \
0      0  https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1      1  https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2      2  https://m.media-amazon.com/images/M/MV5BMTMxNT...   

               Series_Title Released_Year Certificate  Runtime  \
0  The Shawshank Redemption          1994           A  142 min   
1             The Godfather          1972           A  175 min   
2           The Dark Knight          2008          UA  152 min   

                  Genre  IMDB_Rating  \
0                 Drama          9.3   
1          Crime, Drama          9.2   
2  Action, Crime, Drama          9.0   

                                            Overview  Meta_score  \
0  Two imprisoned men bond over a number of years...        80.0   
1  An organized crime dynasty's aging patriarch t...       100.0   
2  When the menace known as the Joker wreaks havo...        84.0   

               Director           Star1           Star2          Star3  \
0        Frank Darabont     Tim Robbins  Morgan Freeman     Bob Gunton   
1  Francis Ford Coppola   Marlon Brando       Al Pacino     James Caan   
2     Christopher Nolan  Christian Bale    Heath Ledger  Aaron Eckhart   

            Star4  No_of_Votes        Gross  
0  William Sadler      2343110   28,341,469  
1    Diane Keaton      1620367  134,966,411  
2   Michael Caine      2303232  534,858,444

In [ ]:
#escrever/salvar em xml:
dados_imdb.to_xml('filmes_imdb.xml')

**07 Para saber mais: o que é XML?**
XML é uma linguagem de marcação que permite criar e descrever dados de forma estruturada e padronizada. Ela é composta por linhas de comandos que usam tags para definir os elementos e os atributos dos dados.

A estrutura básica de um documento XML é formada por:

Uma declaração inicial que indica a versão, o encoding e o tipo do documento.
Um elemento raiz que contém todos os outros elementos do documento.
Elementos filhos que podem ter outros elementos ou conteúdo de texto dentro deles.
Atributos que fornecem informações adicionais sobre os elementos.
Um exemplo simples de um documento XML é:

xml
<?xml version="1.0" encoding="UTF-8"?>
<animais>
  <animal nome="Mel" tipo="cachorro" cor="marrom"/>
  <animal nome="Vick" tipo="gato" cor="branco"/>
</animais>Copiar código
Neste exemplo, <animais> é o elemento raiz, <animal> é um elemento filho com três atributos (nome, tipo e cor) e Mel e Vick são os conteúdos de texto dos elementos.

Podemos encontrar mais informações sobre a estrutura XML na página: XML para iniciantes - Suporte da Microsoft.


**Aula 05 Trabalhando com banco de dados**

In [ ]:
#SQL  = linguagem de consulta estruturada
# para trabalharmos com esses dados precisaremos realizar as seguintes importações:
# importaremos a biblioteca sqlalchemy
import sqlalchemy
from sqlalchemy import create_engine, MetaData, Table, inspect
# create_engine : fará a conexao
# MetaData: classe que representa os metadados
# Table : Classe que permite trabalhar com tabelas
# inspect: permite inspecionar os dados

In [ ]:
engine = create_engine('sqlite:///:memory:')

In [ ]:
import pandas as pd

In [ ]:
dados_clientes = pd.read_csv('/content/clientes_banco.csv')
dados_clientes

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008804     32   Ensino superior  União-estável              2.0   
1         5008805     32   Ensino superior  União-estável              2.0   
2         5008806     58      Ensino médio         Casado              2.0   
3         5008808     52      Ensino médio       Solteiro              1.0   
4         5008809     52      Ensino médio       Solteiro              1.0   
...           ...    ...               ...            ...              ...   
31695     5092120     59      Ensino médio         Casado              2.0   
31696     5092121     59      Ensino médio         Casado              2.0   
31697     5092122     59      Ensino médio         Casado              2.0   
31698     5092123     59      Ensino médio         Casado              2.0   
31699     5092124     59      Ensino médio            Cas              NaN   

        Categoria_de_renda   Ocupacao  Anos_empregado  Rendimento_anual  \
0                Empregado      Outro            12.0          427500.0   
1                Empregado      Outro            12.0          427500.0   
2                Empregado  Segurança             3.0          112500.0   
3      Associado comercial     Vendas             8.0          270000.0   
4      Associado comercial     Vendas             8.0          270000.0   
...                    ...        ...             ...               ...   
31695          Pensionista      Outro             0.0          193500.0   
31696          Pensionista      Outro             0.0          193500.0   
31697          Pensionista      Outro             0.0          193500.0   
31698          Pensionista      Outro             0.0          193500.0   
31699                  NaN        NaN             NaN               NaN   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0       Apartamento alugado  
2            1.0  Casa/apartamento próprio  
3            0.0  Casa/apartamento próprio  
4            0.0  Casa/apartamento próprio  
...          ...                       ...  
31695        0.0  Casa/apartamento próprio  
31696        0.0  Casa/apartamento próprio  
31697        0.0  Casa/apartamento próprio  
31698        0.0  Casa/apartamento próprio  
31699        NaN                       NaN  

[31700 rows x 11 columns]

In [ ]:
#colocar os dados dentro do banco de dados:
dados_clientes.to_sql('clientes_2', engine, index = False)
#clientes = nome da tabela
#engine = é o motor, variável criada anteriormente
#index = false = Para não criar uma nova coluna de índice



31700

In [ ]:
inspector = inspect(engine) #verificar o que tem dentro do banco de dados
print(inspector.get_table_names()) # get_table_names() = função que devolve s nomes das tabelas que está dentro do banco de dados

['clientes_2']


Bancos de dados são sistemas que permitem o armazenamento, organização e recuperação de informações de forma estruturada e eficiente. Eles são amplamente utilizados em muitas áreas, desde gerenciamento de negócios à pesquisa científica. Um banco de dados bem projetado pode melhorar significativamente a eficiência e a precisão das operações de uma organização.

Existem bancos de dados relacionais e não-relacionais. Os bancos de dados relacionais representam e armazenam dados em tabelas. Já os não-relacionais, também conhecidos como bancos de dados NoSQL (Not Only SQL - Não SQL), usam uma variedade de estruturas de dados, como documentos, grafos ou pares de chave-valor.

Python oferece vários pacotes e bibliotecas para trabalhar com bancos de dados, incluindo SQLite, MySQL, PostgreSQL, Oracle, MongoDB, entre outros. Um dos pacotes mais comuns usados para trabalhar com bancos de dados relacionais em Python é o pacote sqlite3 que oferece suporte a bancos de dados SQLite. Este banco de dados é leve e incorporado que não exige um servidor separado para ser executado que já vem nativamente instalado no Google Colab.

Para trabalhar com esse banco podemos utilizar a SQLAlchemy, uma biblioteca de mapeamento objeto-relacional (ORM), que possibilita interagir com bancos de dados relacionais usando código Python. Ela fornece uma camada de abstração que permite aos desenvolvedores trabalhar com objetos Python em vez de lidar diretamente com as complexidades da linguagem SQL (Structured Query Language - Linguagem de consulta estruturada).

Uma das principais vantagens do uso de SQLAchemy é a capacidade de criar código mais legível e fácil de manter. Com SQLAlchemy, as operações do banco de dados são executadas usando métodos em objetos Python, tornando o código mais claro e menos propenso a erros.

Além disso, a SQLAlchemy oferece suporte a consultas complexas em bancos de dados, permitindo que pessoas desenvolvedoras extraiam facilmente informações relevantes de grandes conjuntos de dados. Isso é especialmente útil em aplicações que precisam lidar com grandes quantidades de dados.

Você pode encontrar mais detalhes sobre a biblioteca SQLAchemy na documentação.

https://www.sqlalchemy.org/

**Para saber mais: cláusulas SQL**
SQL (Structured Query Language - Linguagem de consulta estruturada) é uma linguagem de consulta utilizada em bancos de dados relacionais para inserir, atualizar, consultar e gerenciar dados.

Para fazer tudo isso,existem as cláusulas SQL, que são componentes fundamentais das instruções SQL, permitindo especificar detalhes sobre como a consulta ou a operação de banco de dados devem ser executadas. As cláusulas são usadas para filtrar, classificar, agrupar e limitar os resultados das consultas.

As instruções SQL podem ser compostas por uma ou mais cláusulas que fornecem informações adicionais sobre o que a consulta deve fazer. As cláusulas mais comuns são:

SELECT: especifica quais colunas devem ser selecionadas na consulta.
FROM: especifica as tabelas do banco de dados que devem ser consultadas.
WHERE: filtra os resultados da consulta com base em uma ou mais condições especificadas.
ORDER BY: classifica os resultados da consulta em ordem crescente ou decrescente com base em uma ou mais colunas.
GROUP BY: agrupa os resultados da consulta com base em uma ou mais colunas.
LIMIT: limita o número de linhas retornadas pelos resultados da consulta.
Temos um exemplo simples de sintaxe SQL que usa o comando SELECT para consultar dados de uma tabela de colaboradores(as) em um banco de dados:

SELECT nome, sobrenome, salario
FROM colaboradores
WHERE departamento = 'vendas'Copiar código
Nessa instrução SQL, a cláusula SELECT é usada para especificar as colunas que se deseja consultar da tabela "colaboradores", incluindo "nome", "sobrenome" e "salario". A cláusula FROM é usada para especificar a tabela que se deseja consultar, que é "colaboradores" neste caso.

A cláusula WHERE é usada para filtrar os resultados da consulta com base em uma condição especificada. Neste exemplo, a condição é departamento = 'vendas', o que significa que a consulta retornará apenas as pessoas colaboradoras que trabalham no departamento de vendas.

In [ ]:
#utilizamos a funçaõ query para localizar algo dentro da base de dados:

query = 'SELECT * FROM clientes_2 WHERE Categoria_de_renda = "Empregado"'
# O asterico indica que quero consultar todas as colunas

In [ ]:
empregados = pd.read_sql(query, engine)

In [ ]:
empregados

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008804     32   Ensino superior  União-estável              2.0   
1         5008805     32   Ensino superior  União-estável              2.0   
2         5008806     58      Ensino médio         Casado              2.0   
3         5008815     46   Ensino superior         Casado              2.0   
4         5112956     46   Ensino superior         Casado              2.0   
...           ...    ...               ...            ...              ...   
15660     5092104     45      Ensino médio       Solteiro              2.0   
15661     5092105     45      Ensino médio       Solteiro              2.0   
15662     5092106     45      Ensino médio       Solteiro              2.0   
15663     5092108     45      Ensino médio       Solteiro              2.0   
15664     5092109     45      Ensino médio       Solteiro              2.0   

      Categoria_de_renda       Ocupacao  Anos_empregado  Rendimento_anual  \
0              Empregado          Outro            12.0          427500.0   
1              Empregado          Outro            12.0          427500.0   
2              Empregado      Segurança             3.0          112500.0   
3              Empregado  Contabilidade             2.0          270000.0   
4              Empregado  Contabilidade             2.0          270000.0   
...                  ...            ...             ...               ...   
15660          Empregado       Medicina            25.0          135000.0   
15661          Empregado       Medicina            25.0          135000.0   
15662          Empregado       Medicina            25.0          135000.0   
15663          Empregado       Medicina            25.0          135000.0   
15664          Empregado       Medicina            25.0          135000.0   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0       Apartamento alugado  
2            1.0  Casa/apartamento próprio  
3            1.0  Casa/apartamento próprio  
4            1.0  Casa/apartamento próprio  
...          ...                       ...  
15660        0.0        Habitação pública   
15661        0.0        Habitação pública   
15662        0.0        Habitação pública   
15663        0.0        Habitação pública   
15664        0.0        Habitação pública   

[15665 rows x 11 columns]

In [ ]:
empregados.to_sql('empregados', con=engine, index = False)
# con=engine = parâmetro para especificar a conexão
# Aqui fizemos uma consulta

15665

In [ ]:
# para ler a tabela inteira

pd.read_sql_table('empregados',engine) #coloca o nome da tabela

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008804     32   Ensino superior  União-estável              2.0   
1         5008805     32   Ensino superior  União-estável              2.0   
2         5008806     58      Ensino médio         Casado              2.0   
3         5008815     46   Ensino superior         Casado              2.0   
4         5112956     46   Ensino superior         Casado              2.0   
...           ...    ...               ...            ...              ...   
15660     5092104     45      Ensino médio       Solteiro              2.0   
15661     5092105     45      Ensino médio       Solteiro              2.0   
15662     5092106     45      Ensino médio       Solteiro              2.0   
15663     5092108     45      Ensino médio       Solteiro              2.0   
15664     5092109     45      Ensino médio       Solteiro              2.0   

      Categoria_de_renda       Ocupacao  Anos_empregado  Rendimento_anual  \
0              Empregado          Outro            12.0          427500.0   
1              Empregado          Outro            12.0          427500.0   
2              Empregado      Segurança             3.0          112500.0   
3              Empregado  Contabilidade             2.0          270000.0   
4              Empregado  Contabilidade             2.0          270000.0   
...                  ...            ...             ...               ...   
15660          Empregado       Medicina            25.0          135000.0   
15661          Empregado       Medicina            25.0          135000.0   
15662          Empregado       Medicina            25.0          135000.0   
15663          Empregado       Medicina            25.0          135000.0   
15664          Empregado       Medicina            25.0          135000.0   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0       Apartamento alugado  
2            1.0  Casa/apartamento próprio  
3            1.0  Casa/apartamento próprio  
4            1.0  Casa/apartamento próprio  
...          ...                       ...  
15660        0.0        Habitação pública   
15661        0.0        Habitação pública   
15662        0.0        Habitação pública   
15663        0.0        Habitação pública   
15664        0.0        Habitação pública   

[15665 rows x 11 columns]

In [ ]:
pd.read_sql_table('empregados',engine,columns=['ID_Cliente','Grau_escolaridade','Rendimento_anual' ]) #escolhendo as colunas que desejo visualizar

ID_Cliente Grau_escolaridade  Rendimento_anual
0         5008804   Ensino superior          427500.0
1         5008805   Ensino superior          427500.0
2         5008806      Ensino médio          112500.0
3         5008815   Ensino superior          270000.0
4         5112956   Ensino superior          270000.0
...           ...               ...               ...
15660     5092104      Ensino médio          135000.0
15661     5092105      Ensino médio          135000.0
15662     5092106      Ensino médio          135000.0
15663     5092108      Ensino médio          135000.0
15664     5092109      Ensino médio          135000.0

[15665 rows x 3 columns]

**Antes de atualizar o banco de dados…**

Antes de atualizar o banco de dados…
No próximo vídeo, vamos aprender como realizar algumas modificações nos dados. Porém, é importante dizer que foram implementadas algumas atualizações recentes na biblioteca que usamos, a SQLAlchemy.

Para garantir que você consiga chegar nos mesmos resultados apresentados no vídeo, precisamos abordar os métodos atualizados para interação com o banco de dados.

Após ler todos os registros da tabela clientes, é fundamental iniciar importando a função text do SQLAlchemy. Esta função é essencial para criar e executar expressões SQL literais dentro do seu código Python. Ao utilizar text, você pode inserir diretamente instruções SQL que o SQLAlchemy compreenderá e executará:

from sqlalchemy import text

Para deletar um registro na tabela clientes, agora utilizamos a seguinte estrutura:

query = 'DELETE FROM clientes WHERE ID_Cliente=5008804'
with engine.connect() as conn:
    result = conn.execute(text(query))
    conn.commit()  

Vamos quebrar o código em partes para entender melhor cada etapa:

Definição da consulta SQL: Primeiro, definimos o que queremos fazer com o banco de dados em uma string de consulta SQL, armazenada na variável query. Aqui, o comando DELETE é usado para remover o registro de um cliente específico, identificado pelo ID_Cliente=5008804.

Estabelecendo conexão com o banco de dados: Usamos with engine.connect() as conn para abrir uma conexão com o banco. O with é muito útil porque gerencia automaticamente a abertura e o fechamento da conexão. Isso significa que, após o bloco de código ser executado, a conexão será fechada automaticamente, evitando que recursos sejam desperdiçados ou bloqueios no banco de dados se prolonguem.

Executando a consulta: Dentro do bloco with, executamos a consulta com conn.execute(text(query)). A função text() é usada aqui para garantir que a string de consulta seja interpretada como SQL puro.

Confirmar a transação: Por fim, chamamos conn.commit() para confirmar a transação. Isso é crucial porque, em muitos bancos de dados, as alterações feitas (como deletar um registro) não são salvas permanentemente até que sejam confirmadas explicitamente.

Este processo garante que o registro seja removido de forma segura e eficiente, minimizando a possibilidade de erros ou problemas de segurança.

De maneira similar, para atualizar informações de um registro, como o grau de escolaridade de um cliente, você pode proceder da seguinte forma:

query = 'UPDATE clientes SET Grau_escolaridade="Ensino superior" WHERE ID_Cliente=5008808'
with engine.connect() as conn:
    result = conn.execute(text(query))
    conn.commit()  

Aqui, a conexão é estabelecida novamente, a consulta de atualização é executada para alterar o grau de escolaridade do cliente com ID 5008808 para "Ensino superior", e a transação é confirmada.

Estes passos são cruciais para manipular dados de forma segura e eficaz, utilizando as funcionalidades mais recentes do SQLAlchemy. Seguir estas instruções ajudará a assegurar que você obtenha resultados precisos e consistentes, conforme demonstrado no vídeo.

Então, dê play no vídeo a seguir, experimente os códigos e veja como é eficaz trabalhar com essa ferramenta atualizada!


In [ ]:
query = 'SELECT * FROM clientes_2'

In [ ]:
pd.read_sql(query, engine)

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008805     32   Ensino superior  União-estável              2.0   
1         5008806     58      Ensino médio         Casado              2.0   
2         5008808     52   Ensino superior       Solteiro              1.0   
3         5008809     52      Ensino médio       Solteiro              1.0   
4         5008810     52      Ensino médio       Solteiro              1.0   
...           ...    ...               ...            ...              ...   
31694     5092120     59      Ensino médio         Casado              2.0   
31695     5092121     59      Ensino médio         Casado              2.0   
31696     5092122     59      Ensino médio         Casado              2.0   
31697     5092123     59      Ensino médio         Casado              2.0   
31698     5092124     59      Ensino médio            Cas              NaN   

        Categoria_de_renda   Ocupacao  Anos_empregado  Rendimento_anual  \
0                Empregado      Outro            12.0          427500.0   
1                Empregado  Segurança             3.0          112500.0   
2      Associado comercial     Vendas             8.0          270000.0   
3      Associado comercial     Vendas             8.0          270000.0   
4      Associado comercial     Vendas             8.0          270000.0   
...                    ...        ...             ...               ...   
31694          Pensionista      Outro             0.0          193500.0   
31695          Pensionista      Outro             0.0          193500.0   
31696          Pensionista      Outro             0.0          193500.0   
31697          Pensionista      Outro             0.0          193500.0   
31698                 None       None             NaN               NaN   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0  Casa/apartamento próprio  
2            0.0  Casa/apartamento próprio  
3            0.0  Casa/apartamento próprio  
4            0.0  Casa/apartamento próprio  
...          ...                       ...  
31694        0.0  Casa/apartamento próprio  
31695        0.0  Casa/apartamento próprio  
31696        0.0  Casa/apartamento próprio  
31697        0.0  Casa/apartamento próprio  
31698        NaN                      None  

[31699 rows x 11 columns]

In [ ]:
from sqlalchemy import text

In [ ]:
# utilizo a query para procurar  a informação
query = 'DELETE FROM clientes_2 WHERE ID_Cliente=5008804'
with engine.connect() as conn: #with serve para garantir que a conexao com o banco de dados seja fechada corretamente após a eexecução da consulta
    result = conn.execute(text(query))
    conn.commit()

In [ ]:
pd.read_sql_table('clientes_2', engine)

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008805     32   Ensino superior  União-estável              2.0   
1         5008806     58      Ensino médio         Casado              2.0   
2         5008808     52   Ensino superior       Solteiro              1.0   
3         5008809     52      Ensino médio       Solteiro              1.0   
4         5008810     52      Ensino médio       Solteiro              1.0   
...           ...    ...               ...            ...              ...   
31694     5092120     59      Ensino médio         Casado              2.0   
31695     5092121     59      Ensino médio         Casado              2.0   
31696     5092122     59      Ensino médio         Casado              2.0   
31697     5092123     59      Ensino médio         Casado              2.0   
31698     5092124     59      Ensino médio            Cas              NaN   

        Categoria_de_renda   Ocupacao  Anos_empregado  Rendimento_anual  \
0                Empregado      Outro            12.0          427500.0   
1                Empregado  Segurança             3.0          112500.0   
2      Associado comercial     Vendas             8.0          270000.0   
3      Associado comercial     Vendas             8.0          270000.0   
4      Associado comercial     Vendas             8.0          270000.0   
...                    ...        ...             ...               ...   
31694          Pensionista      Outro             0.0          193500.0   
31695          Pensionista      Outro             0.0          193500.0   
31696          Pensionista      Outro             0.0          193500.0   
31697          Pensionista      Outro             0.0          193500.0   
31698                 None       None             NaN               NaN   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0  Casa/apartamento próprio  
2            0.0  Casa/apartamento próprio  
3            0.0  Casa/apartamento próprio  
4            0.0  Casa/apartamento próprio  
...          ...                       ...  
31694        0.0  Casa/apartamento próprio  
31695        0.0  Casa/apartamento próprio  
31696        0.0  Casa/apartamento próprio  
31697        0.0  Casa/apartamento próprio  
31698        NaN                      None  

[31699 rows x 11 columns]

In [ ]:
# para atualizarmos uma informação iremos utilizar a cláusula UPDATE

query = 'UPDATE clientes_2 SET Grau_escolaridade="Ensino superior" WHERE ID_Cliente=5008808'
with engine.connect() as conn:
    result = conn.execute(text(query))
    conn.commit()

In [ ]:
pd.read_sql_table('clientes_2', engine)

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0         5008805     32   Ensino superior  União-estável              2.0   
1         5008806     58      Ensino médio         Casado              2.0   
2         5008808     52   Ensino superior       Solteiro              1.0   
3         5008809     52      Ensino médio       Solteiro              1.0   
4         5008810     52      Ensino médio       Solteiro              1.0   
...           ...    ...               ...            ...              ...   
31694     5092120     59      Ensino médio         Casado              2.0   
31695     5092121     59      Ensino médio         Casado              2.0   
31696     5092122     59      Ensino médio         Casado              2.0   
31697     5092123     59      Ensino médio         Casado              2.0   
31698     5092124     59      Ensino médio            Cas              NaN   

        Categoria_de_renda   Ocupacao  Anos_empregado  Rendimento_anual  \
0                Empregado      Outro            12.0          427500.0   
1                Empregado  Segurança             3.0          112500.0   
2      Associado comercial     Vendas             8.0          270000.0   
3      Associado comercial     Vendas             8.0          270000.0   
4      Associado comercial     Vendas             8.0          270000.0   
...                    ...        ...             ...               ...   
31694          Pensionista      Outro             0.0          193500.0   
31695          Pensionista      Outro             0.0          193500.0   
31696          Pensionista      Outro             0.0          193500.0   
31697          Pensionista      Outro             0.0          193500.0   
31698                 None       None             NaN               NaN   

       Tem_carro                   Moradia  
0            1.0       Apartamento alugado  
1            1.0  Casa/apartamento próprio  
2            0.0  Casa/apartamento próprio  
3            0.0  Casa/apartamento próprio  
4            0.0  Casa/apartamento próprio  
...          ...                       ...  
31694        0.0  Casa/apartamento próprio  
31695        0.0  Casa/apartamento próprio  
31696        0.0  Casa/apartamento próprio  
31697        0.0  Casa/apartamento próprio  
31698        NaN                      None  

[31699 rows x 11 columns]